File: data_structures/5_kotlin/e_binarySearchTrees.kt
class TreeNode<T : Comparable<T>>(val value: T) {
    var left: TreeNode<T>? = null
    var right: TreeNode<T>? = null
}

class BinarySearchTree<T : Comparable<T>> {
    private var root: TreeNode<T>? = null

    // MARK: - Insertion

    fun insert(value: T) {
        root = insertRec(root, value)
    }

    private fun insertRec(node: TreeNode<T>?, value: T): TreeNode<T> {
        if (node == null) {
            return TreeNode(value)
        }

        if (value < node.value) {
            node.left = insertRec(node.left, value)
        } else if (value > node.value) {
            node.right = insertRec(node.right, value)
        }

        return node
    }

    // MARK: - Deletion

    fun delete(value: T) {
        root = deleteRec(root, value)
    }

    private fun deleteRec(node: TreeNode<T>?, value: T): TreeNode<T>? {
        if (node == null) {
            return null
        }

        if (value < node.value) {
            node.left = deleteRec(node.left, value)
        } else if (value > node.value) {
            node.right = deleteRec(node.right, value)
        } else {
            if (node.left == null) {
                return node.right
            } else if (node.right == null) {
                return node.left
            }

            val minRight = findMin(node.right!!)
            node.value = minRight.value
            node.right = deleteRec(node.right, minRight.value)
        }

        return node
    }

    // Helper method to find the minimum value node in a subtree
    private fun findMin(node: TreeNode<T>): TreeNode<T> {
        var current = node
        while (current.left != null) {
            current = current.left!!
        }
        return current
    }

    // MARK: - Search

    fun search(value: T): Boolean {
        return searchRec(root, value)
    }

    private fun searchRec(node: TreeNode<T>?, value: T): Boolean {
        if (node == null) {
            return false
        }

        if (value == node.value) {
            return true
        } else if (value < node.value) {
            return searchRec(node.left, value)
        } else {
            return searchRec(node.right, value)
        }
    }

    // MARK: - Traversal

    fun inorderTraversal(): List<T> {
        val result = mutableListOf<T>()
        inorderTraversalRec(root, result)
        return result
    }

    private fun inorderTraversalRec(node: TreeNode<T>?, result: MutableList<T>) {
        if (node != null) {
            inorderTraversalRec(node.left, result)
            result.add(node.value)
            inorderTraversalRec(node.right, result)
        }
    }

    // ... Add more BST operations as needed ...
}

// Example usage
fun main() {
    val bst = BinarySearchTree<Int>()

    bst.insert(5)
    bst.insert(3)
    bst.insert(7)
    bst.insert(1)
    bst.insert(4)

    println("In-order traversal: ${bst.inorderTraversal()}")
    println("Search 4: ${bst.search(4)}")
    println("Search 6: ${bst.search(6)}")

    bst.delete(3)

    println("In-order traversal after deletion: ${bst.inorderTraversal()}")
}

File: data_structures/3_python/e_binarySearchTrees.py
class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


class BinarySearchTree:
    def __init__(self):
        self.root = None

    # MARK: - Insertion

    def insert(self, value):
        self.root = self._insert_rec(self.root, value)

    def _insert_rec(self, node, value):
        if node is None:
            return TreeNode(value)

        if value < node.value:
            node.left = self._insert_rec(node.left, value)
        elif value > node.value:
            node.right = self._insert_rec(node.right, value)

        return node

    # MARK: - Deletion

    def delete(self, value):
        self.root = self._delete_rec(self.root, value)

    def _delete_rec(self, node, value):
        if node is None:
            return None

        if value < node.value:
            node.left = self._delete_rec(node.left, value)
        elif value > node.value:
            node.right = self._delete_rec(node.right, value)
        else:
            if node.left is None:
                return node.right
            elif node.right is None:
                return node.left

            min_right = self._find_min(node.right)
            node.value = min_right.value
            node.right = self._delete_rec(node.right, min_right.value)

        return node

    def _find_min(self, node):
        current = node
        while current.left:
            current = current.left
        return current

    # MARK: - Search

    def search(self, value):
        return self._search_rec(self.root, value)

    def _search_rec(self, node, value):
        if node is None:
            return False

        if value == node.value:
            return True
        elif value < node.value:
            return self._search_rec(node.left, value)
        else:
            return self._search_rec(node.right, value)

    # MARK: - Traversal

    def inorder_traversal(self):
        result = []
        self._inorder_traversal_rec(self.root, result)
        return result

    def _inorder_traversal_rec(self, node, result):
        if node:
            self._inorder_traversal_rec(node.left, result)
            result.append(node.value)
            self._inorder_traversal_rec(node.right, result)

    # ... Add more BST operations as needed ...


# Example usage
if __name__ == "__main__":
    bst = BinarySearchTree()

    bst.insert(5)
    bst.insert(3)
    bst.insert(7)
    bst.insert(1)
    bst.insert(4)

    print("In-order traversal:", bst.inorder_traversal())
    print("Search 4:", bst.search(4))
    print("Search 6:", bst.search(6))

    bst.delete(3)

    print("In-order traversal after deletion:", bst.inorder_traversal())
